src/osp/hierarchical_bitset.h

```diff
@@ -0,0 +1,563 @@
+
+#endif
+
+/**
+ * @brief Divide and round up
+ */
+template<typename NUM_T, typename DENOM_T>
+constexpr decltype(auto) div_ceil(NUM_T num, DENOM_T denom) noexcept
+{
+    return (num / denom) + (num % denom != 0);
+}
+
+template<typename INT_T>
+constexpr bool bit_test(INT_T block, int bit)
+{
+    return (block & (INT_T(1) << bit) ) != 0;
+}
+
+template<typename INT_T>
+constexpr void copy_bits(INT_T const* pSrc, INT_T* pDest, size_t bits) noexcept
+{
+    while (bits >= sizeof(INT_T) * 8)
+    {
+        *pDest = *pSrc;
+        pSrc ++;
+        pDest ++;
+        bits -= sizeof(INT_T) * 8;
+    }
+
+    if (bits != 0)
+    {
+        INT_T const mask = (~INT_T(0)) << bits;
+
+        *pDest &= mask;
+        *pDest |= ( (~mask) & (*pSrc) );
+    }
+}
+
+template<typename INT_T>
+constexpr void set_bits(INT_T* pDest, size_t bits) noexcept
+{
+    while (bits >= sizeof(INT_T) * 8)
+    {
+        *pDest = ~INT_T(0);
+        pDest ++;
+        bits -= sizeof(INT_T) * 8;
+    }
+
+    if (bits != 0)
+    {
+        *pDest |= ~( (~INT_T(0)) << bits );
+    }
+}
+
+//-----------------------------------------------------------------------------
+
+/**
+ * @brief A bitset featuring hierarchical rows for fast iteration over set bits
+ *
+ * For example, when using 8-bit block sizes, the structure looks like this:
+ *
+ * Row 1: 00000101
+ * Row 0: 00000011 00000000 00001100 00000000 00000000
+ *
+ * Size: 40 bits, determined by Row 0
+ * Bits set: { 0, 1, 18, 19 }
+ *
+ * note: Bits are displayed in msb->lsb, but arrays are left->right
+ *
+ * * Row 1's bit 0 and bit 2 means Row 0's block 0 and block 2 is non-zero
+ * * Row 1's last 3 bits are always zero
+ * * Container supports sizes that aren't aligned with block sizes, which means
+ *   the last bits of Row 0's last block may be ignored
+ */
+template <typename BLOCK_INT_T = uint64_t>
+class HierarchicalBitset
+{
+    using row_index_t = size_t;
+
+    static constexpr int smc_blockSize = sizeof(BLOCK_INT_T) * 8;
+
+    // Max size of the top row.
+    // A higher number means functions like take() must search the top level for
+    // non-zero blocks, but 1 less row to manage. Unlikely this will do anything
+    static constexpr int smc_topLevelMaxBlocks = 1;
+
+    // Max bits representable = smc_blockSize^smc_maxRows
+    // Due to its exponential nature, this number doesn't need to be that high
+    static constexpr int smc_maxRows = 8;
+
+    struct RowBit { row_index_t m_block; int m_bit; };
+
+public:
+
+    struct Row { size_t m_offset; size_t m_size; };
+    using rows_t = std::array<Row, smc_maxRows>;
+
+    HierarchicalBitset() = default;
+
+    /**
+     * @brief Construct with size
+     *
+     * @param size [in] Size in bits
+     * @param fill [in] Initialize with 1 bits if true
+     */
+    HierarchicalBitset(size_t size, bool fill = false)
+     : m_size{size}
+     , m_topLevel{0}
+    {
+        m_blockCount = calc_blocks_recurse(size, 0, m_topLevel, m_rows);
```

Comment:
These can still be part of the initialization block, so long as they are in the same order that they are declared.